### DISCUSSION TITLE
Should the default distance method for cylindrical color spaces use polar coordinates?

### DISCUSSION THREAD

**Discussion by ilikescience** [color-js/color.js#493] [Ideas]
When when working in cylindrical color spaces like hsl, hsluv, lch, etc (essentially anything with a "hue" component that goes from 0 → 360) the default "euclidian" distance $\sqrt{(x_1 - x_2)^2 + (y_1 - y_2)^2 + (z_1 - z_2)^2}$ doesn't make much sense. For example, the euclidian distance between hsl(0, 50%, 50%) and hsl(360, 50%, 50%) is calculated as $\sqrt{(0 - 360)^2 + (0.5 - 0.5)^2 + (0.5 - 0.5)^2} = 360$ , even though hsl(0, 50%, 50%) and hsl(360, 50%, 50%) are the same color!

Instead, I think we should first convert the polar coordinates to cartesian ones before calculating the euclidean distance:

```math
\displaylines{x=r*cos(\Theta) \\ y=r*sin(\Theta) \\ z=z}
```

where r and z are the normal/orthogonal coordinates — in hsl, for example, they would be s and l — and θ is the radial coordinate, in radians — in hsl, this would be $h*\frac{\pi}{180}$.

This would mean that the euclidian distance between hsl(0, 50%, 50%) and hsl(360, 50%, 50%) would be correctly calculated as 0.

I think the key concern here is that for each cylindrical space, we'd need to indicate which coordinate is the radial one and which are the normal ones. But with that information, euclidian distances would be much more logical.

**Comment by facelessuser** [comment]
I'm not sure I understand the concern. The default ∆E is 76, and it uses Lab, not a cylindrical space, to perform the distance calculation. https://github.com/color-js/color.js/blob/main/src/deltaE/deltaE76.js#L6.

```js
> new Color('red').deltaE(new Color('blue'))
184.01904862099684
> new Color('red').to('hsluv').deltaE(new Color('blue').to('hsluv'))
184.0190486209965
```

**Reply by facelessuser** [reply]
I'm sorry, I didn't read fully. You are specifically requesting it to use Cylindrical values, but that is not what 76 does, per its specification. This would not be ∆E76 if it did anything other than what it was designed to do.

**Reply by ilikescience** [reply]
I'm specifically referring to the `distance()` method, not the `deltaE()` method.

```js
const color1 = new Color('hsl', [0, 0, 0]);
const color2 = new Color('hsl', [360, 0, 0]);
color1.distance(color2, 'hsl') // → 360
```

**Reply by facelessuser** [reply]
I think maybe I did read correctly. So maybe you are talking about the `distance` function which is pure Euclidean, and in that case, its default is still `lab`: https://github.com/color-js/color.js/blob/main/src/distance.js#L7. You have to manually tell it to use HSL or HSLuv, or some other cylindrical colors space. If you don't want cylindrical coordinates to be used, then I would suggest using a cartesian equivalent and not tell it to use a cylindrical space.

**Reply by facelessuser** [reply]
We can see it defaults to using `lab` and does give zero:

```js
> new Color('hsl', [0, 0, 0]).distance(new Color('hsl', [360, 0, 0]))
0
```

**Reply by ilikescience** [reply]
Right, if we measure the distance in a cartesian space like the default lab or srgb, the distance is given as 0, which is correct. But like in my example in this thread, if we measure in a cylindrical space, we're getting an incorrect value — the `distance()` function assumes all color spaces are cartesian.

**Reply by facelessuser** [reply]
Do you have a particular use case where you feel it is more helpful to use euclidean distance in an unsupported cylindrical space instead of just using a more appropriate cartesian coordinate system? Or are you mainly suggesting that _if_ people do so, we should try to provide more sane defaults?

Normally, if I want distance in HSL, I would use sRGB which represents its gamut, I wouldn't use HSL directly. I could see value in maybe throwing a warning or error if a cylindrical space is used. I guess the alternative would be to convert to a cartesian system to try and return more sane results. 

In order to do so, we'd need a way to determine in every color space where hue and some colorfulness/saturation/chroma component lies. Not all LCh spaces place hue at the end (HCT as an example), etc. Is it worth all the effort to provide such behaviour, or is it better to just guide users to not try and do distance in cylindrical spaces?

**Reply by ilikescience** [reply]
There's a few cases where I like working in a cylindrical space — specifically, hsluv, though I've also been using okhsl a bit lately too. Those spaces hit a sweet spot between computational simplicity and approximate perceptual uniformity; I can go into more detail if you'd like.

In hsluv, euclidian distance agrees pretty closely to deltaE2000, so you can use it as a heuristic instead of converting to another color space (ie converting to lab to do deltaE2000). 

Today, if I'm working in a cylindrical space and want to avoid doing conversions, the `distance()` method will give me the incorrect euclidian distance.

As for implementation: each color space declares the type for each of its coords; could we check that to determine if we're in a cylindrical space, and if so, which coordinate is angular?

**Reply by facelessuser** [reply]
@ilikescience, I understand that working in hsluv, and okhsl can be great, I am not trying to discourage that, but why not use the appropriate space for distancing, Luv and Oklab respectively?

```js
> let c1 = new Color('hsluv', [0, 0, 0])
undefined
> let c2 = new Color('hsluv', [360, 0, 0])
undefined
> c1.distance(c2, 'luv')
0
> let c3 = new Color('okhsl', [0, 0, 0])
undefined
> let c4 = new Color('okhsl', [360, 0, 0])
undefined
> c3.distance(c4, 'oklab')
0
```

Hopefully, this makes things a bit more clear at what I'm driving at. Is it more helpful to you to get distancing within the actual HSL cylindrical space? Or do you actually want the Luv or Oklab distancing which actually gives you the results closer to ∆E2000? Distancing in the distorted HSL space will not give you what you want I imagine.

**Reply by ilikescience** [reply]
I don't think it's correct to say that oklab is more appropriate for making distance measurements than okhsl; they're useful in their own rights. In my case, okhsl has some really useful properties. For example, I can calculate a "maximally different color," because there's an easily derived "furthest point" in okhsl space, which is guaranteed to be in the srgb gamut. 

Before okhsl and hsluv were added to colorjs.io, I was supplying my own method for doing these kinds of distance calculations in those spaces. That's ok, and I'm sure folks have opinions on whether or not colorjs.io should support this use case. For now, maybe it's sufficient to focus on the fact that if the `distance()` function is meant to be a euclidian distance it's currently not behaving correctly for cylindrical spaces. If we don't want to support euclidian distances for cylindrical spaces, maybe the library should throw an error?

**Reply by facelessuser** [reply]
> I don't think it's correct to say that oklab is more appropriate for making distance measurements than okhsl; they're useful in their own rights. In my case, okhsl has some really useful properties. For example, I can calculate a "maximally different color," because there's an easily derived "furthest point" in okhsl space, which is guaranteed to be in the srgb gamut.

So we aren't talking about perceptual distance, that is what I was getting at, and it is fine if that is not what you are looking for. You specifically want distancing in okhsl and hsluv. That clears that up, and that is okay.

How would you propose HWB would get handled? I think spaces like this may still be an exception, but that is probably okay.

As it seems you truly do want distancing in these cylindrical spaces, I will defer to others for an actual decision on the matter.

I think the problem could be solved if a decision is made to do so, but there may or may not need to be  a little framework put in place to generically identify saturation/chroma/colorfulness and hue in spaces. Hue is easy enough as I believe they identify as angles. Maybe you can safely assume the center component is always saturation/chroma/colorfulness. I'm not aware of a specific space that doesn't follow this convention, but I can confidently say that currently supported spaces follow this convention. So maybe no additional infrastructure is needed?

**Comment by ilikescience** [comment]
> So we aren't talking about perceptual distance, that is what I was getting at, and it is fine if that is not what you are looking for. You specifically want distancing in okhsl and hsluv. That clears that up, and that is okay.

Yup, I'm just talking about the geometric distance in the color space. To be pedantic, this is a kind of perceptual distance, in the same way that geometric distance in jzazbz or lab is meant to be perceptually meaningful.

> How would you propose HWB would get handled?

Since it's cylindrical, it should be handled the same way; distance should be calculated using cylindrical coordinates.

> but there may or may not need to be a little framework put in place to generically identify saturation/chroma/colorfulness and hue in spaces

It looks like there's already a method on ColorSpaces that identifies spaces as being polar or not! eg,

```js
> Color.space.srgb.isPolar
false
> Color.space.hsluv.isPolar
true
```

Behind the scenes it's just checking if a given coord has is of type `angle`:

```js
this.coords[id].type === "angle"
```

So, the distance method could just sort coords into angular and cartesian, then do the distance calculation appropriately.

The "right" way to do this would be to also allow for spherical color spaces, ie spaces where two out of the three coordinates are angular — while I don't think any color spaces fit this description today, it's not outside the realm of possibility. It'd be _extremely_ right to also allow for arbitrary numbers of coordinates (2, 3, 4, ...n), and I'm sure there's a generalized formula for distance in n-dimensions (where some of them can be angular), but I'm not a mathematician so I think we'll have to punt on that one :)

**Reply by facelessuser** [reply]
> Since it's cylindrical, it should be handled the same way; distance should be calculated using cylindrical coordinates.

How would you calculate that? It's not the same as HSL, HSV, or LCh spaces.

**Reply by facelessuser** [reply]
I guess you would use W as Chroma, I forget that is how I plot it in 3D. I guess you could use either white or black, I guess it only matters on how you want to present it.

**Reply by facelessuser** [reply]
> this.coords[id].type === "angle"

I am aware of this, I was more thinking about chroma. It seems in currently provided spaces it is always in the middle, I am unaware of any that do not follow this convention, but if that were to ever not be true, you would need a more explicit way to target chroma/saturation/colorfulness.

Anyway, now that I understand what you are asking, this is more a question of whether Color.js wants to support this, I will leave that to the decision makers.

**Reply by ilikescience** [reply]
> you would need a more explicit way to target chroma/saturation/colorfulness

I don't think that's necessary; could you say more? In a cylindrical space, it doesn't matter which cartesian coordinate is which ... like you mentioned with w/b in hwb, you can swap the two and get the same distance.

**Reply by facelessuser** [reply]
>  it doesn't matter which cartesian coordinate is which ... like you mentioned with w/b in hwb, you can swap the two and get the same distance.

Let's not confuse things, it does matter whether you take W or B when calculating cartesian coordinates for HWB.

[Example](https://facelessuser.github.io/coloraide/playground/?code=import%20math%0Afrom%20coloraide%20import%20algebra%20as%20alg%0A%0Adef%20distance_hwb%28color1%2C%20color2%2C%20swap%3DFalse%29%3A%0A%20%20%20%20h1%2C%20w1%2C%20b1%20%3D%20Color%28color1%29%2econvert%28%27hwb%27%29%2ecoords%28nans%3DFalse%29%0A%20%20%20%20h2%2C%20w2%2C%20b2%20%3D%20Color%28color2%29%2econvert%28%27hwb%27%29%2ecoords%28nans%3DFalse%29%0A%20%20%20%20dx%20%3D%20%28b2%20-%20b1%29%20if%20not%20swap%20else%20%28w2%20-%20w1%29%0A%20%20%20%20y1%2C%20z1%20%3D%20alg%2epolar_to_rect%28w1%20if%20not%20swap%20else%20b1%2C%20h1%29%0A%20%20%20%20y2%2C%20z2%20%3D%20alg%2epolar_to_rect%28w2%20if%20not%20swap%20else%20b2%2C%20h2%29%0A%20%20%20%20dy%20%3D%20y2%20-%20y1%0A%20%20%20%20dz%20%3D%20z2%20-%20z1%0A%20%20%20%20return%20math%2esqrt%28dx%20%2a%2a%202%20%2B%20dy%20%2a%2a%202%20%2B%20dz%20%2a%2a%202%29%0A%0Adistance_hwb%28%27lightblue%27%2C%20%27darkgreen%27%29%0Adistance_hwb%28%27lightblue%27%2C%20%27darkgreen%27%2C%20swap%3DTrue%29%0A%0A%0A%20%20%20%20).

<img width="1356" alt="Screenshot 2024-03-25 at 9 48 57 AM" src="https://github.com/color-js/color.js/assets/1055125/23fc4bed-b107-41af-a210-b9547438e6b3">

**Reply by ilikescience** [reply]
Yes, sorry, you're absolutely right, these things are pretty unintuitive!

Ok, given that the choice of radial dimension is meaningful, you would need a way of correctly assigning the coordinate. That does seem more involved, as it requires adding extra information to each color space.

**Reply by facelessuser** [reply]
Yep, hence my slight concern about locating chroma. Assuming such things were in place, and it was decided that this project wanted to support polar Euclidean distance, I think it would be fine. HWB handling, which I imagine is the only "special" case that would need consideration, would have to be decided then: pick a convention, or throw an exception.

**Reply by ilikescience** [reply]
I was just doing some research into HWB and it looks like in the original proposal, W is akin to Saturation and B is akin to Value. So, W would be the radial dimension.

http://alvyray.com/Papers/CG/HWB_JGTv208.pdf

![Screenshot 2024-03-25 at 12 17 21@2x](https://github.com/color-js/color.js/assets/1400729/c2389666-d72d-4908-9bc0-21a03d0138a6)

**Reply by facelessuser** [reply]
That was my memory, or at least what I settled on. That's how I plot them:

[Example](https://facelessuser.github.io/coloraide/demos/3d_models.html?space=hwb&gamut=srgb&edges=false&aspect=false&ortho=false).

<img width="639" alt="Screenshot 2024-03-25 at 10 35 38 AM" src="https://github.com/color-js/color.js/assets/1055125/f28b45d5-1f08-4057-9c27-e5570db9e9c0">

**Comment by ilikescience** [comment]
@LeaVerou @svgeesus what do y'all think? Should the `.distance()` method consider whether a space is polar or not? Is it worth the additional effort to indicate how polar spaces are oriented?

**Reply by LeaVerou** [reply]
Absolutely!! The question is who will implement it 😅 Any chance you may be up for it? 🙏🏼
